### PR TITLE
Feature/modularize wasm engine

### PR DESCRIPTION
**Pull request by iamyulong**
This PR modularizes wasm validation and execution with the following traits:
- `ScryptoModule` represents an invokable Scrypto module that following Scrypto call ABI;
- `ScryptoRuntime` represents the runtime to support the execution of a Scrypto module;
- `ScryptoWasmValidator`, `ScryptoWasmInstrumenter`  and `ScryptoWasmExecutor` defines the traits for validating, instrumenting and executing Scrypto modules;

The default implementation is provided by `WasmiEngine` in the `wasmi` module.

### CODE REVIEW

**Review by talekhinezh — APPROVED** [review]
Just a few minor comments

File: radix-engine/src/engine/call_frame.rs [hunk 1]
```diff
@@ -752,12 +672,12 @@ impl<'r, 'l, L: ReadableSubstateStore> Process<'r, 'l, L> {
...
         );
         if let Some(auth_zone) = &self.auth_zone {
-            process.caller_auth_zone = Option::Some(auth_zone);
+            frame.caller_auth_zone = Option::Some(auth_zone);
         }
 
         // invoke the main function
         let (result, received_buckets, received_proofs, received_vaults) =
-            process.run(Some(snode_ref), snode, call_data)?;
+            frame.run(Some(snode_ref), snode, call_data)?;
```
[the remark this answers is not in the document]
Comment (iamyulong) [reply radix-engine/src/engine/call_frame.rs:680]: Agreed. It's a bit weird. Maybe `start` or `main`?

File: radix-engine/src/model/package.rs [hunk 2]
```diff
@@ -1,30 +1,20 @@
...
+
+use crate::engine::SystemApi;
+use crate::wasm::*;
 
 /// A collection of blueprints, compiled and published as a single unit.
 #[derive(Debug, Clone, TypeId, Encode, Decode)]
 pub struct Package {
     code: Vec<u8>,
-    blueprints: HashMap<String, Type>,
+    blueprints: HashMap<String, (Type, Vec<Function>, Vec<Method>)>,
```
[the remark this answers is not in the document]
Comment (iamyulong) [reply radix-engine/src/model/package.rs:17]: Haha, didn't want to refactor the invoke ABI logic so just did this. :)

File: radix-engine/src/model/package.rs [hunk 3]
```diff
@@ -38,52 +28,33 @@ pub enum PackageError {
...
-            .iter()
-            .filter(|(name, val)| {
-                name.ends_with("_abi") && name.len() > 4 && matches!(val, ExternVal::Func(_))
-            })
-            .map(|(name, _)| name.to_string())
+        let mut wasm_engine = WasmiEngine::new();
+        wasm_engine.validate(&code)?;
+
+        let module = wasm_engine.instantiate(&code);
+        let exports: Vec<String> = module
```
[the remark this answers is not in the document]
Comment (iamyulong) [reply radix-engine/src/model/package.rs:35]: Yes, I can do this.

File: radix-engine/src/wasm/constants.rs [hunk 4]
```diff
@@ -0,0 +1,8 @@
+pub const ENGINE_FUNCTION_INDEX: usize = 0;
+pub const ENGINE_FUNCTION_NAME: &str = "radix_engine";
+
+pub const EXPORT_ENV: &str = "env";
+
+pub const EXPORT_MEMORY: &str = "memory";
+pub const EXPORT_SCRYPTO_ALLOC: &str = "scrypto_alloc";
+pub const EXPORT_SCRYPTO_FREE: &str = "scrypto_free";
```
[the remark this answers is not in the document]
Comment (iamyulong) [reply radix-engine/src/wasm/constants.rs:8]: Yeah, I'd like to remove them as well. Will need to research vendor-agnostic direct memory access.

**Pull request merged**